### PR TITLE
Added a lint rule for palette references

### LINKED ISSUE #8689: Add a lint pass to verify traitinfo palette references.
<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/23993314-add-a-lint-pass-to-verify-traitinfo-palette-references?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Mailaender**
Closes #8689

### CODE REVIEW

File: OpenRA.Game/Traits/Player/PlayerColorPalette.cs [hunk 1]
```diff
@@ -17,13 +17,17 @@ namespace OpenRA.Traits
 	public class PlayerColorPaletteInfo : ITraitInfo
 	{
 		[Desc("The name of the palette to base off.")]
-		public readonly string BasePalette = null;
+		[PaletteReference] public readonly string BasePalette = null;
+
 		[Desc("The prefix for the resulting player palettes")]
-		public readonly string BaseName = "player";
+		[PaletteDefinition] public readonly string BaseName = "player";
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/PlayerColorPalette.cs]: This is a player palette, so the BaseName by itself is not a valid palette name.

File: OpenRA.Game/Traits/Player/PlayerHighlightPalette.cs [hunk 2]
```diff
@@ -19,7 +19,7 @@ namespace OpenRA.Traits
 	public class PlayerHighlightPaletteInfo : ITraitInfo
 	{
 		[Desc("The prefix for the resulting player palettes")]
-		public readonly string BaseName = "highlight";
+		[PaletteDefinition] public readonly string BaseName = "highlight";
```
Comment (pchote) [inline OpenRA.Game/Traits/Player/PlayerHighlightPalette.cs]: Likewise here.

File: OpenRA.Mods.Common/Traits/Buildings/RallyPoint.cs [hunk 3]
```diff
@@ -18,7 +18,7 @@ namespace OpenRA.Mods.Common.Traits
 	public class RallyPointInfo : ITraitInfo
 	{
 		public readonly CVec RallyPoint = new CVec(1, 3);
-		public readonly string IndicatorPalettePrefix = "player";
+		[PaletteReference] public readonly string IndicatorPalettePrefix = "player";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Buildings/RallyPoint.cs]: This references (I think) a player palette, so regular palette names are not valid here.

File: OpenRA.Mods.Common/Traits/Render/RenderSprites.cs [hunk 4]
```diff
@@ -34,10 +34,10 @@ public class RenderSpritesInfo : IRenderActorPreviewInfo, ITraitInfo
 		public readonly Dictionary<string, string> RaceImages = null;
 
 		[Desc("Custom palette name")]
-		public readonly string Palette = null;
+		[PaletteReference] public readonly string Palette = null;
 
 		[Desc("Custom PlayerColorPalette: BaseName")]
-		public readonly string PlayerPalette = "player";
+		[PaletteReference] public readonly string PlayerPalette = "player";
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Render/RenderSprites.cs]: ... and so on for these other player palette references (no need to spam up the discussion listing all of them).

**Comment by pchote** [comment]
I suggest adding overloads for the palette reference that let you specify as a bool whether or not it is a player palette, or a string that points at the bool field that determines at runtime whether it is a player palette.

**Comment by Mailaender** [comment]
Rebased and avoided player prefixed palettes due to time and complexity constrains.

**Comment by abcdefg30** [comment]
Looks good to me. :+1: / :white_check_mark:

File: OpenRA.Mods.Common/Lint/CheckPalettes.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+			}
+
+			foreach (var weaponInfo in rules.Weapons)
+			{
+				var projectileInfo = weaponInfo.Value.Projectile;
+				if (projectileInfo == null)
+					continue;
+
+				var fields = projectileInfo.GetType().GetFields();
+				foreach (var field in fields)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Lint/CheckPalettes.cs]: Why actually not `fields.Where(...)` like above?

File: OpenRA.Mods.Common/Lint/CheckPalettes.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	class CheckPalettes : ILintPass
+	{
+		public void Run(Action<string> emitError, Action<string> emitWarning, Map map)
+		{
+			if (map != null && !map.RuleDefinitions.Any())
+				return;
+
+			var rules = map == null ? Game.ModData.DefaultRules : map.Rules;
+
+			var palettes = GetPalettes(emitError, rules);
```
Comment (penev92) [inline OpenRA.Mods.Common/Lint/CheckPalettes.cs]: `.ToList()` here, please.

File: OpenRA.Mods.Common/Lint/CheckPalettes.cs (hunk 5 above)
Comment (penev92) [inline OpenRA.Mods.Common/Lint/CheckPalettes.cs]: Yes please, for consistency and easier reading.

**Comment by penev92** [comment]
:+1:  with the above two requests.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by Mailaender** [comment]
Updated.

**Comment by penev92** [comment]
Thanks

**Comment by penev92** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/5dd08158d19c77ccb0703e4677aa593240b05cfa...3911aeca577bf3cb3ece4c28c49b6897772fd3a5)

**Pull request merged**